src/functions.c

```diff
@@ -0,0 +1,446 @@
+    if (!fi) {
+        addReplyError(c, "Function not found");
+        return;
+    }
+
+    engineFunctionFree(fi, functions_ctx);
+    forceCommandPropagation(c, PROPAGATE_REPL | PROPAGATE_AOF);
+    addReply(c, shared.ok);
+}
+
+void functionsKillCommand(client *c) {
+    scriptKill(c, 0);
+}
+
+static void fcallCommandGeneric(client *c, int ro) {
+    robj *function_name = c->argv[1];
+    functionInfo *fi = dictFetchValue(functions_ctx->functions, function_name->ptr);
+    if (!fi) {
+        addReplyError(c, "Function not found");
+        return;
+    }
+    engine *engine = fi->ei->engine;
+
+    long long numkeys;
+    /* Get the number of arguments that are keys */
+    if (getLongLongFromObject(c->argv[2], &numkeys) != C_OK) {
+        addReplyError(c, "Bad number of keys provided");
+        return;
+    }
+    if (numkeys > (c->argc - 3)) {
+        addReplyError(c, "Number of keys can't be greater than number of args");
+        return;
+    } else if (numkeys < 0) {
+        addReplyError(c, "Number of keys can't be negative");
+        return;
+    }
+
+    scriptRunCtx run_ctx;
+
+    scriptPrepareForRun(&run_ctx, fi->ei->c, c, fi->name);
+    if (ro) {
+        run_ctx.flags |= SCRIPT_READ_ONLY;
+    }
+    engine->call(&run_ctx, engine->engine_ctx, fi->function, c->argv + 3, numkeys,
+                 c->argv + 3 + numkeys, c->argc - 3 - numkeys);
+    scriptResetRun(&run_ctx);
+}
+
+/*
+ * FCALL <FUNCTION NAME> nkeys <key1 .. keyn> <arg1 .. argn>
+ */
+void fcallCommand(client *c) {
+    fcallCommandGeneric(c, 0);
+}
+
+/*
+ * FCALL_RO <FUNCTION NAME> nkeys <key1 .. keyn> <arg1 .. argn>
+ */
+void fcallCommandReadOnly(client *c) {
+    fcallCommandGeneric(c, 1);
+}
+
+void functionsHelpCommand(client *c) {
+    const char *help[] = {
+"CREATE <ENGINE NAME> <FUNCTION NAME> [REPLACE] [DESC <FUNCTION DESCRIPTION>] <FUNCTION CODE>",
+"    Create a new function with the given function name and code.",
+"DELETE <FUNCTION NAME>",
+"    Delete the given function.",
+"INFO",
+"    For each function, print the following information about the function:",
+"    * Function name",
+"    * The engine used to run the function",
+"    * Function description",
+"KILL",
+"    Kill the current running function.",
+NULL };
+    addReplyHelp(c, help);
+}
+
+/* Compile and save the given function, return C_OK on success and C_ERR on failure.
+ * In case on failure the err out param is set with relevant error message */
+int functionsCreateWithFunctionCtx(sds function_name,sds engine_name, sds desc, sds code,
+                                   int replace, sds* err, functionsCtx *functions) {
+    engineInfo *ei = dictFetchValue(engines, engine_name);
+    if (!ei) {
+        *err = sdsnew("Engine not found");
+        return C_ERR;
+    }
+    engine *engine = ei->engine;
+
+    functionInfo *fi = dictFetchValue(functions->functions, function_name);
+    if (fi) {
+        if (replace) {
+            engineFunctionFree(fi, functions);
```

Comment:
Should free `fi` after `engine->create` successfully?

Reply:
No, we should only free it if it already exists and we replacing it.

Reply:
I mean that when we create a function that fails, we shouldn't delete the old one.

```sh
$ function create lua test replace 'return "hello"'
$ function info
1) 1) "name"
   2) "test"
   3) "engine"
   4) "LUA"
   5) "description"
   6) (nil)
$ function create lua test replace 'error'
(error) ERR Error compiling function: user_function:1: '=' expected near '<eof>'
$ function info
(empty array)
```

Expectation

```sh
$ function create lua test replace 'return "hello"'
$ function info
1) 1) "name"
   2) "test"
   3) "engine"
   4) "LUA"
   5) "description"
   6) (nil)
$ function create lua test replace 'error'
(error) ERR Error compiling function: user_function:1: '=' expected near '<eof>'
$ function info
1) 1) "name"
   2) "test"
   3) "engine"
   4) "LUA"
   5) "description"
   6) (nil)
```

Reply:
Ohh, nice catch, thank you.

Reply:
Added a commit that fixes it.

Reply:
@sundb please let me know if you agree with the fix and if I can resolve this.

Reply:
@MeirShpilraien It's OK.